File: MechanicsCore/src/main/java/me/deecaad/core/file/SerializerException.java
package me.deecaad.core.file;

import me.deecaad.core.utils.Debugger;
import me.deecaad.core.utils.EnumUtil;
import me.deecaad.core.utils.LogLevel;
import me.deecaad.core.utils.StringUtil;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.LinkedList;

public class SerializerException extends Exception {

    private String serializerName;
    private String[] messages;
    private String location;

    public SerializerException(String serializerName, String[] messages, String location) {
        this.serializerName = serializerName;
        this.messages = messages;
        this.location = location;
    }

    /**
     * @param serializer The serializer that generated the exception.
     * @param messages   The messages telling the user the error and how to fix it.
     * @param location   The file + path location to the issue {@link StringUtil#foundAt(File, String)}
     */
    public SerializerException(@NotNull Serializer<?> serializer, String[] messages, @NotNull String location) {
        this.messages = messages;
        this.location = location;
        this.serializerName = serializer.getName();
    }

    public String getSerializerName() {
        return serializerName;
    }

    public void setSerializerName(String serializerName) {
        this.serializerName = serializerName;
    }

    public String[] getMessages() {
        return messages.clone();
    }

    public String getLocation() {
        return location;
    }

    public void setLocation(String location) {
        this.location = location;
    }

    public void log(Debugger debug) {
        log(debug, LogLevel.ERROR);
    }

    public void log(Debugger debug, LogLevel level) {
        LinkedList<String> collected = new LinkedList<>();

        collected.add("A mistake was found in your configurations when making '" + serializerName + "'");
        collected.addAll(Arrays.asList(messages));
        collected.add(location);
        collected.add(""); // Add an empty string for blank line between errors

        debug.log(level, collected.toArray(new String[0]));
    }

    public SerializerException addMessage(boolean condition, String message) {
        if (condition)
            addMessage(message);

        return this;
    }

    public SerializerException addMessage(String message) {
        String[] copy = new String[messages.length + 1];
        System.arraycopy(messages, 0, copy, 0, messages.length);
        copy[messages.length] = message;
        this.messages = copy;

        return this;
    }


    public static String forValue(Object value) {
        return "Found value: " + value;
    }

    public static <T extends Enum<T>> String didYouMean(String input, Class<T> enumClass) {
        return didYouMean(input, EnumUtil.getOptions(enumClass));
    }

    public static String didYouMean(String input, Iterable<String> options) {
        String expected = StringUtil.didYouMean(input, options);
        return "Did you mean to use '" + expected + "' instead of '" + input + "'?";
    }

    public static String examples(String... examples) {
        StringBuilder builder = new StringBuilder("'");

        for (String str : examples)
            builder.append(str).append("', ");

        builder.setLength(builder.length() - 2);

        return "Example values: " + builder;
    }

    public static String possibleValues(Iterable<String> options, String actual, int count) {

        // We want to know how many elements are in the iterable. We don't care
        // about performance. We need this to tell the user how many options there
        // actually are. This is important since we may only show 5 or 6 options,
        // but it is important for the USER to know that there are more than 5 or 6.
        ArrayList<String> arr = new ArrayList<>();
        options.forEach(arr::add);
        int[] table = StringUtil.mapToCharTable(actual);

        // Sort the list based on what is most similar to 'actual'.
        arr.sort((a, b) -> {
            int[] localA = StringUtil.mapToCharTable(a);
            int[] localB = StringUtil.mapToCharTable(b);

            int differenceA = Math.abs(actual.length() - a.length());
            int differenceB = Math.abs(actual.length() - b.length());

            for (int i = 0; i < table.length; i++) {
                differenceA += Math.abs(table[i] - localA[i]);
                differenceB += Math.abs(table[i] - localB[i]);
            }

            return Integer.compare(differenceA, differenceB);
        });

        count = Math.min(arr.size(), count);
        StringBuilder builder = new StringBuilder("Showing ");

        // Writes either 'All' or a fraction like '4/32'
        if (count == arr.size())
            builder.append("All");
        else
            builder.append(count).append('/').append(arr.size());

        builder.append(" Options:");

        // Append some possible values
        Iterator<String> iterator = arr.iterator();
        while (iterator.hasNext() && count-- > 0) {
            String next = iterator.next();
            builder.append(" '").append(next).append("'");
        }

        return builder.toString();
    }
}

File: MechanicsCore/src/main/kotlin/me/deecaad/core/file/SerializerException.kt
package me.deecaad.core.file

import me.deecaad.core.utils.Debugger
import me.deecaad.core.utils.EnumUtil
import me.deecaad.core.utils.LogLevel
import me.deecaad.core.utils.StringUtil
import java.io.File
import java.util.stream.Stream
import kotlin.math.abs

/**
 * An exception that is thrown when a mistake is found in the config file.
 * These mistakes are typically caused by the user, and not the plugin itself.
 *
 * Serializer exceptions are designed to be caught by the plugin during the
 * deserialization process, and then logged to the console. This way, users
 * see very "pretty" error messages instead of ugly stack traces. This makes
 * it more possible for people to fix their own errors.
 *
 * @param location A human-readable location of the mistake (file location + path, generally)
 * @param messages A list of messages that describe the mistake in more detail
 */
open class SerializerException(
    private val location: String,
    private val messages: MutableList<String> = mutableListOf()
) : Exception() {

    @JvmOverloads
    fun log(debug: Debugger, level: LogLevel = LogLevel.ERROR) {
        val fullError = mutableListOf<String>()
        fullError.add("A mistake was found in your config!")
        fullError.addAll(messages)
        fullError.add(location)
        fullError.add("")  // An empty line to separate the error from the rest of the log

        debug.log(level, fullError)
    }

    class Builder {
        private var location: String? = null
        private val messages: MutableList<String> = mutableListOf()

        fun location(
            file: File,
            path: String,
            index: Int? = null
        ): Builder {
            this.location = when {
                index != null -> "Located in file '$file' at '$path' (The ${StringUtil.ordinal(index)} list item)"
                else -> "Located in file '$file' at '$path'"
            }
            return this
        }

        fun locationRaw(location: String): Builder {
            this.location = location
            return this
        }

        fun example(exampleValue: String): Builder {
            messages.add("Example value: $exampleValue")
            return this
        }

        fun addMessage(message: String): Builder {
            messages.add(message)
            return this
        }

        fun didYouMean(actual: String, options: Iterable<String>) {
            val expected = StringUtil.didYouMean(actual, options)
            messages.add("Did you mean to use '$expected' instead of '$actual'?")
        }

        fun possibleValues(actual: String, options: Iterable<String>, count: Int): Builder {
            val arr = options.toList()
            val actualTable = StringUtil.toCharTable(actual)

            val sortedArr = arr.sortedWith { a, b ->
                val aTable = StringUtil.toCharTable(a)
                val bTable = StringUtil.toCharTable(b)

                var diffA = abs(actual.length - a.length)
                var diffB = abs(actual.length - b.length)

                for (i in actualTable.indices) {
                    diffA += abs(actualTable[i] - aTable[i])
                    diffB += abs(actualTable[i] - bTable[i])
                }

                diffA.compareTo(diffB)
            }

            val limitedCount = minOf(sortedArr.size, count)
            val builder = StringBuilder("Showing ")

            if (limitedCount == sortedArr.size) {
                builder.append("All")
            } else {
                builder.append("$limitedCount/${sortedArr.size}")
            }

            builder.append(" Options:")

            if (limitedCount > 0) {
                builder.append(sortedArr.take(limitedCount).joinToString("") { " '$it'" })
            }

            messages.add(builder.toString())
            return this
        }

        fun buildInvalidRange(actual: Int, min: Int?, max: Int?): SerializerException {
            if (location == null) {
                throw IllegalStateException("Location must be set before calling buildInvalidRange")
            }

            messages.add("Invalid range! Expected a value between ${min ?: "-∞"} and ${max ?: "∞"}")
            messages.add("Found value: $actual")

            return SerializerException(location!!, messages)
        }

        fun buildInvalidRange(actual: Double, min: Double?, max: Double?): SerializerException {
            if (location == null) {
                throw IllegalStateException("Location must be set before calling buildInvalidRange")
            }

            messages.add("Invalid range! Expected a value between ${min ?: "-∞"} and ${max ?: "∞"}")
            messages.add("Found value: $actual")

            return SerializerException(location!!, messages)
        }

        fun buildInvalidRegistryOption(input: String, registry: org.bukkit.Registry<*>): SerializerException {
            return buildInvalidOption(input, registry.map { it.key.key })
        }

        fun <T : Enum<T>> buildInvalidEnumOption(input: String, enumClass: Class<T>): SerializerException {
            return buildInvalidOption(input, EnumUtil.getOptions(enumClass))
        }

        fun buildInvalidOption(input: String, options: Iterable<String>): SerializerException {
            if (location == null) {
                throw IllegalStateException("Location must be set before calling buildInvalidOption")
            }

            messages.add("Unknown value '$input'")
            didYouMean(input, options)
            possibleValues(input, options, 5)

            return SerializerException(location!!, messages)
        }

        fun buildMissingRequiredKey(missingKey: String): SerializerException {
            if (location == null) {
                throw IllegalStateException("Location must be set before calling buildMissingRequiredKey")
            }

            messages.add("Missing required key '$missingKey'")
            messages.add("Make sure you spell it correctly (case sensitive)")

            return SerializerException(location!!, messages)
        }

        fun buildInvalidType(expectedTyped: String, actualValue: Any) : SerializerException {
            if (location == null) {
                throw IllegalStateException("Location must be set before calling buildInvalidType")
            }

            messages.add("Invalid type! Expected a $expectedTyped")
            messages.add("Found value: $actualValue")

            return SerializerException(location!!, messages)
        }

        fun build() : SerializerException {
            if (location == null) {
                throw IllegalStateException("Location must be set before calling build")
            }

            return SerializerException(location!!, messages)
        }
    }

    companion object {
        @JvmStatic
        fun builder(): Builder {
            return Builder()
        }
    }
}